src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,221 @@
+    public T set(T value,int index) {
+        validateIndex(index);
+        Node target = getNodeByIndex(index);
+        T result = target.value;
+        target.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        Node target = getNodeByIndex(index);
+        T result = target.value;
+        if (size == 1) {
+            tail = null;
+            head = null;
+        } else if (size == 2) {
+            if (target.value.equals(head.value)) {
+                head = tail;
+            } else {
+                tail = head;
+            }
+            wire(head, tail);
+        } else if (index == 0) {
+            head = target.next;
+            head.prev = head;
+        } else if (index == size - 1) {
+            tail = target.prev;
+            tail.next = null;
+        } else {
+            wire(target.prev, target.next);
+        }
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T t) {
-        return false;
+        Node target = getNodeByValue(t);
+        if (target == null) {
+            return false;
+        }
+
+        if (size == 1) {
+            tail = null;
+            head = null;
+        } else if (size == 2) {
+            if (target.value == head.value || target.value.equals(head.value)) {
+                head = tail;
+            } else {
+                tail = head;
+            }
+            wire(head, tail);
+            release(target);
+        } else if (head.value == t || head.value.equals(t)) {
+            head = target.next;
+            head.prev = head;
+            release(target);
+        } else {
+            if (target != tail) {
+                wire(target.prev, target.next);
+            } else if (tail.value == t || tail.value.equals(t)) {
+                tail = target.prev;
+                tail.next = null;
+            }
+            release(target);
+        }
+        size--;
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (size == 0);
+    }
+
+    private void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException(index);
+        }
+    }
+
+    private void release(Node node) {
+        node.prev = null;
+        node.next = null;
+    }
+
+    private void wire(Node node1, Node node2) {
+        node1.next = node2;
+        node2.prev = node1;
+    }
+
+    private Node getNodeByIndex(int index) {
+        Node result = null;
+
+        if (index < size / 2) {
+            result = head;
+            for (int i = 0; i <= index; i++) {
+                if (i == index) {
+                    return result;
+                }
+                result = result.next;
+            }
+        } else {
```

Comment:
```suggestion
            for (int i = 0; i < index; i++) {
                result = result.next;
            }
        } else {
```
we don't need if check, because we gonna exit loop whenever  i < index is not valid anymore